Json.More/EnumStringConverter.cs

```diff
@@ -151,10 +151,10 @@ public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions
 
 		private static void EnsureMap()
 		{
-			if (_readValues != null) return;
+			if (_writeValues != null) return;
 			lock (_lock)
 			{
-				if (_readValues != null) return;
+				if (_writeValues != null) return;
```

Comment:
I'm not sure that this would make a difference.  Both are set within this lock.  Checking one for null should be equivalent to checking the other.

Reply:
Thanks for reviewing my PR. I've added unit tests in `SchemaGenerationTests` to validate the behaviour when `RequiredAttribute` and `JsonPropertyNameAttribute` are set on a property.

In regards to the locking mechanism in `EnumStringConverter.EnsureMap`, the race condition only occurs when the `EnumStringConverter.Write` method is called (serialization to JSON). This race condition is very timing sensitive because it requires the subsequent thread to call `EnumStringConverter.EnsureMap` when `_readValues` is non-null and `_writeValues` is null. Due to this timing sensitivity I am unable to write a unit test to reproduce it, but I have validated on my end that my unit tests running in parallel are no longer failing with this change.

Here is an example of how this condition could be hit:

1. **Thread_A** calls `JsonSerializer.Serialize` on an object that contains an enum property that will be serialized with `EnumStringConverter`.
2. **Thread_B** calls `JsonSerializer.Serialize` on an object (different from step 1) that contains an enum property (same type as step 1) that will be serialized with `EnumStringConverter`.
3. **Thread_A** calls `EnumStringConverter.Write` and then calls `EnumStringConverter.EnsureMap`. The current state of `_readValues` is null, `_writeValues` is null, and `_lock` is unlocked. Therefore **Thread_A** obtain the lock.
4. **Thread_A** sets `map` ([Lines 159-166](https://github.com/gregsdennis/json-everything/blob/master/Json.More/EnumStringConverter.cs#L159-L166)).
5. **Thread_A** sets `_readValues` ([Line 167](https://github.com/gregsdennis/json-everything/blob/master/Json.More/EnumStringConverter.cs#L167)).
6. **Thread_B** calls `EnumStringConverter.Write` and then calls `EnumStringConverter.EnsureMap`. The current state of `_readValues` is non-null, `_writeValues` is null, and `_lock` is locked. Therefore **Thread_B** will return to `EnumStringConverter.Write` ([Line 154](https://github.com/gregsdennis/json-everything/blob/master/Json.More/EnumStringConverter.cs#L154)).
7. **Thread_B** calls `WriteValues.ContainsKey` ([Line 118](https://github.com/gregsdennis/json-everything/blob/master/Json.More/EnumStringConverter.cs#L118)) which results in calling `EnumStringConverter.EnsureMap` again (via the `WriteValues` getter method). The current state of the system has not changed, therefore **Thread_B** will return to the `WriteValues` getter method ([Line 154](https://github.com/gregsdennis/json-everything/blob/master/Json.More/EnumStringConverter.cs#L154)).
8. **Thread_B** throws a `System.NullReferenceException` when calling `WriteValues.ContainsKey` because `WriteValues` (and `_writeValues`) is null.
9. **Thread_A** sets `_writeValues` ([Line 167](https://github.com/gregsdennis/json-everything/blob/master/Json.More/EnumStringConverter.cs#L167)), and then releases the lock.
10. **Thread_A** continues with its execution without issues.

Reply:
Nice analysis. I think I'd be more comfortable with just checking both, then.  If both are non-null, then return early; otherwise go through the lock.  It might be overkill, but I prefer the redundancy.

Reply:
Good point. I've updated it to verify that both `_readValues` and `_writeValues` are non-null.